onnxruntime/core/providers/nnapi/nnapi_builtin/nnapi_execution_provider.cc

```diff
@@ -0,0 +1,356 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+#include "nnapi_execution_provider.h"
+
+#include "builders/model_builder.h"
+#include "core/framework/allocatormgr.h"
+#include "core/framework/compute_capability.h"
+#include "core/graph/model.h"
+#include "core/session/onnxruntime_cxx_api.h"
+#include "core/session/inference_session.h"
+
+namespace onnxruntime {
+
+constexpr const char* NNAPI = "Nnapi";
+
+NnapiExecutionProvider::NnapiExecutionProvider()
+    : IExecutionProvider{onnxruntime::kNnapiExecutionProvider} {
+  DeviceAllocatorRegistrationInfo device_info{
+      OrtMemTypeDefault,
+      [](int) { return onnxruntime::make_unique<CPUAllocator>(
+                    onnxruntime::make_unique<OrtMemoryInfo>(
+                        NNAPI, OrtAllocatorType::OrtDeviceAllocator)); },
+      std::numeric_limits<size_t>::max()};
+
+  InsertAllocator(CreateAllocator(device_info));
+
+  DeviceAllocatorRegistrationInfo cpu_memory_info({OrtMemTypeCPUOutput,
+                                                   [](int) { return onnxruntime::make_unique<CPUAllocator>(onnxruntime::make_unique<OrtMemoryInfo>(NNAPI, OrtAllocatorType::OrtDeviceAllocator, OrtDevice(), 0, OrtMemTypeCPUOutput)); },
+                                                   std::numeric_limits<size_t>::max()});
+
+  InsertAllocator(CreateAllocator(cpu_memory_info));
+}
+
+NnapiExecutionProvider::~NnapiExecutionProvider() {}
+
+std::vector<std::unique_ptr<ComputeCapability>>
+NnapiExecutionProvider::GetCapability(const onnxruntime::GraphViewer& graph,
+                                      const std::vector<const KernelRegistry*>& /*kernel_registries*/) const {
+  std::vector<std::unique_ptr<ComputeCapability>> result;
+  // Handle If and Loop operators
+  if (graph.IsSubgraph()) {
+    return result;
+  }
+
+  // Need access to model_path_
+  for (const auto& tensor : graph.GetAllInitializedTensors()) {
+    if (tensor.second->has_data_location() &&
+        tensor.second->data_location() == ONNX_NAMESPACE::TensorProto_DataLocation_EXTERNAL) {
+      LOGS_DEFAULT(WARNING) << "NNAPI: Initializers with external data"
+                               " location are not currently supported";
+      return result;
+    }
+  }
+
```

Comment:
Possibly too strict, although unlikely to be hit. It's only if there's an initializer you need to use that it matters if its data is external. Can note that as a TODO.

Reply:
TODO as #8 in PR descriptions